### PR TITLE
adds peek, push, reset, remove ops benchmarks

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
- also cleanup and renaming of our extension to bencher

more to come in another pr for allocations/mem.

### CODE REVIEW

File: bench/benches/packets.rs [hunk 1]
```diff
@@ -9,39 +9,124 @@ use proptest::prelude::*;
...
+#[nb2::bench(mempool_capacity = 511)]
+fn parse_benchmarks(c: &mut Criterion) {
+    c.bench_function("packets::parse_packets", |b| {
         let s = v4_udp();
-        b.iter_batches(BATCH_SIZE, s, parse_udp_packet)
+        b.prop_iter_batched(BATCH_SIZE, s, parse_packet)
     });
 
-    c.bench_function("packets::deparse_udp_packets", move |b| {
+    c.bench_function("packets::deparse_packets", |b| {
```
Comment (drunkirishcoder) [inline bench/benches/packets.rs]: if we want to compare results of two bench functions, we should maybe use `benchmark_group`. although in this case, parse and deparse can't really compare. maybe should be two different benches.

https://docs.rs/criterion/0.3.0/criterion/struct.Criterion.html#method.benchmark_group

File: core/src/testils/criterion.rs [hunk 2]
```diff
@@ -5,15 +5,22 @@ use std::cmp;
 use std::time::{Duration, Instant};
 
 pub trait BencherExt {
-    fn iter_batches<R, S, O>(&mut self, batch_size: usize, strategy: S, routine: R)
+    fn prop_iter_batched<R, S, O>(&mut self, batch_size: usize, strategy: S, routine: R)
```
Comment (drunkirishcoder) [inline core/src/testils/criterion.rs]: kinda like `iter` in front, maybe `iter_proptest_batched`?

File: core/src/testils/criterion.rs [hunk 3]
```diff
@@ -5,15 +5,22 @@ use std::cmp;
...
     where
         R: FnMut(S::Value) -> O,
         S: Strategy;
 }
 
 impl BencherExt for Bencher<'_> {
-    fn iter_batches<R, S: Strategy, O>(&mut self, batch_size: usize, strategy: S, mut routine: R)
-    where
+    /// Similar to criterion's iter_batched fn, but allows for a batch, of
```
Comment (drunkirishcoder) [inline core/src/testils/criterion.rs]: the doc reads very confusing. maybe a bit simpler like "similar to iter_batched, but use a proptest strategy as the setup to randomly generate inputs".

also, maybe move the `batch_size` to the end to match the argument order of `iter_batched`.

File: bench/benches/packets.rs [hunk 4]
```diff
@@ -9,39 +9,124 @@ use proptest::prelude::*;
...
+            packet.deparse()
+        });
+        b.prop_iter_batched(BATCH_SIZE, s, peek_into_packet)
+    });
+}
+
+#[nb2::bench(mempool_capacity = 511)]
+fn push_benchmarks(c: &mut Criterion) {
+    c.bench_function("packets::push_into_mbuf", |b| {
+        b.iter_batched(
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply bench/benches/packets.rs]: decided to keep this as is.

File: bench/benches/packets.rs [hunk 5]
```diff
@@ -1,47 +1,275 @@
...
+    group.bench_function("packets::multi_peek_udp", |b| {
+        let s = v4_udp();
+        b.iter_proptest_batched(s, multi_peek_udp, BATCH_SIZE)
+    });
+
+    group.finish()
+}
+
+#[nb2::bench(mempool_capacity = 511)]
+fn single_parse_srh_segments_sizes(c: &mut Criterion) {
```
Comment (drunkirishcoder) [inline bench/benches/packets.rs]: very cool.

File: core/src/testils/proptest/strategy.rs [hunk 6]
```diff
@@ -259,12 +260,18 @@ fn srh<E: Debug + Ipv6Packet>(
...
-        .prop_map(move |(packet, segments, segments_left, tag)| {
+        .prop_perturb(move |(packet, segments, segments_left, tag), mut rng| {
             let mut packet = packet.push::<SegmentRouting<E>>().unwrap();
             packet.set_segments(&segments).unwrap();
-            packet.set_segments_left(segments_left);
             packet.set_tag(tag);
             packet.set_next_header(next_header);
+
+            let num_segments = segments.len() as u8;
+            if segments_left > num_segments {
```
Comment (drunkirishcoder) [inline core/src/testils/proptest/strategy.rs]: cool but I feel this is not a good behavior. it's caller's responsibility to make sure data passed in is sane. don't like functions that decide if your input is not good, I will replace it for you with something. if we want this, we should make a version where segments + segments_left are just completely randomly generated.

**Pull request merged**